lib/filesystem.nix

```diff
@@ -54,4 +54,44 @@
       dir + "/${name}"
   ) (builtins.readDir dir));
 
+  # findSuffix: String -> Path -> [ Path ]
+  #
+  # Given a suffix and directory, return a filtered list containing the full
+  # path to all files matching the given suffix recursively.
+  #
+  # Example:
+  #   findSuffix ".h" "${glibc.dev}/include/protocols"
+  #
+  # Returns:
+  #  [
+  #   "/nix/store/p5vg4rdba95snvxd7p0km695jnm8vvwh-glibc-2.35-163-dev/include/protocols/routed.h"
+  #   "/nix/store/p5vg4rdba95snvxd7p0km695jnm8vvwh-glibc-2.35-163-dev/include/protocols/rwhod.h"
+  #   "/nix/store/p5vg4rdba95snvxd7p0km695jnm8vvwh-glibc-2.35-163-dev/include/protocols/talkd.h"
+  #   "/nix/store/p5vg4rdba95snvxd7p0km695jnm8vvwh-glibc-2.35-163-dev/include/protocols/timed.h"
+  # ]
+  findSuffix = suffix: dir: (builtins.filter
```

Comment:
This seems a bit redundant when you could already do `find ".*\\.h"`.

Reply:
See the above resolved conversation https://github.com/NixOS/nixpkgs/pull/188354#discussion_r955759338, I do not believe that it's right to force regexp when suffixes are fine for a lot of cases.

Reply:
I really don't think this warrants two functions.  The regex is trivial.

Reply:
@alyssais I couldn't have been more incorrect about performance. I ran the benchmarks. Also, TIL about `lib.strings.escapeRegex`

```
❯ cat regexp.nix
(import ./. {}).lib.filesystem.find ".+\\.nix" ./.

❯ hyperfine --warmup 5 'nix eval -f ./regexp.nix'
Benchmark 1: nix eval -f ./regexp.nix
  Time (mean ± σ):     433.4 ms ±   7.0 ms    [User: 206.2 ms, System: 230.2 ms]
  Range (min … max):   425.3 ms … 449.4 ms    10 runs

❯ cat suffix.nix
(import ./. {}).lib.filesystem.findSuffix ".nix" ./.

❯ hyperfine --warmup 5 'nix eval -f ./suffix.nix'
Benchmark 1: nix eval -f ./suffix.nix
  Time (mean ± σ):      7.824 s ±  0.125 s    [User: 1.379 s, System: 3.730 s]
  Range (min … max):    7.591 s …  7.987 s    10 runs
```

Reply:
Actually, this was because `findSuffix` was accidentally copying everything into the store. I've applied `toString` to the `findSuffix` function argument, and now it's slightly more performant than the regexp variant.

```
❯ hyperfine --warmup 5 'nix eval -f ./regexp.nix'
Benchmark 1: nix eval -f ./regexp.nix
  Time (mean ± σ):     402.5 ms ±   7.0 ms    [User: 178.8 ms, System: 226.9 ms]
  Range (min … max):   392.1 ms … 414.0 ms    10 runs

❯ hyperfine --warmup 5 'nix eval -f ./suffix.nix'
Benchmark 1: nix eval -f ./suffix.nix
  Time (mean ± σ):     392.1 ms ±   8.6 ms    [User: 161.6 ms, System: 233.6 ms]
  Range (min … max):   381.2 ms … 403.9 ms    10 runs
```

Reply:
fwiw you can go also go `find -name "*.h"`.

Reply:
> In my view, regexp is too complicated, and I don't want it.

I agree, regexes are too complicated, even such a simple one. Requires users to remember that it's `.*` and not `*` like with bash, it requires users to remember to prefix it with `.*`, it requires users to remember to escape the `.` with `\.`, but then actually also escape that within Nix so it becomes `\\.`. Also it requires users to know whether the regex needs to be terminated, should it be `^.*\\.h$` just in case? Not needed for Nix but it's really not super obvious.